tool2schema/schema.py

```diff
@@ -105,108 +113,106 @@ def wrapper(function):
 
 
 class FunctionSchema:
-    """Automatically create a function schema."""
+    """Automatically create a function schema for OpenAI."""
+
+    def __init__(self, f: Callable, schema_type: SchemaType = SchemaType.API):
+        """
+        Initialize FunctionSchema for the given function.
 
-    def __init__(self, f: Callable):
+        :param f: The function to create a schema for;
+        :param schema_type: Type of schema;
+        """
         self.f = f
-        self.schema = FunctionSchema.schema(f)
+        self.schema_type: SchemaType = schema_type
+        self.schema: dict = {}
+        self.parameter_schemas: dict[str, ParameterSchema] = {}
+        self._populate_schema()
 
     def to_json(self, schema_type: SchemaType = SchemaType.API) -> dict:
         """
         Convert schema to JSON.
         :param schema_type: Type of schema to return
         """
         if schema_type == SchemaType.TUNE:
-            return FunctionSchema.schema(self.f, schema_type)["function"]
+            return FunctionSchema(self.f, schema_type).to_json()["function"]
```

Comment:
This works, but it feels like a slight misdirection (or slightly abusing the `SchemaType` enum).

What do you think about populating `self.parameter_schemas` within the constructor but waiting to actually create the json in the `to_json` method? That way you don't need to create an entirely new object (redoing some computation) to create a slightly different JSON format.

Reply:
I agree with you and we can improve it, I did it like this to mimic the existing code:
```python
    def to_json(self, schema_type: SchemaType = SchemaType.API) -> dict:
        if schema_type == SchemaType.TUNE:
            # we invoke the static method and re-compute the json
            return FunctionSchema.schema(self.f, schema_type)["function"]
        return self.schema
```

I will change it to compute the parameters on initialisation and the json dictionary only upon request. Maybe I can imitate the `_add_...` methods of ParameterSchema.

It would also be beneficial because currently we compute the schema dictionary for every function that has a decorator as soon as the decorator is instantiated, while in this way the schema would not be computed until the user invokes `to_json` (also much easier to debug).

Reply:
> I did it like this to mimic the existing code

Oops haha.

Reply:
Checked in, let me know what you think :)

Reply:
Skimmed through it and it looks good! Will have a proper look through when it's ready.